units/src/amount/signed.rs

```diff
@@ -63,7 +63,29 @@ impl SignedAmount {
     pub const MAX: Self = SignedAmount::MAX_MONEY;
 
     /// Constructs a new [`SignedAmount`] with satoshi precision and the given number of satoshis.
-    pub const fn from_sat(satoshi: i64) -> SignedAmount { SignedAmount(satoshi) }
+    ///
+    /// # Errors
+    ///
+    /// On values exceeding [`SignedAmount::MAX`] or less than [`SignedAmount::MIN`]
+    pub const fn from_sat(satoshi: i64) -> Result<SignedAmount, OutOfRangeError> {
+        if satoshi < Self::MIN.0 {
+            Err(OutOfRangeError { is_signed: true, is_greater_than_max: false })
+        } else if satoshi > Self::MAX.0 {
+            Err(OutOfRangeError { is_signed: true, is_greater_than_max: true })
+        } else {
+            Ok(SignedAmount(satoshi))
+        }
+    }
+
+    /// Constructs a new [`SignedAmount`] with satoshi precision and the given number of satoshis.
+    /// Note that use of this function is unsafe since no boundary checking is done.  That is, it's
+    /// possible to create a [`SignedAmount`] that is less than [`SignedAmount::MIN`] or greater
+    /// than [`SignedAmount::MAX`].  Use of [`SignedAmount`] in this way is discouraged and it may
+    /// be deprecated in the future to dis-allow this behavior.  Consider using a different type if
+    /// you need to represent values that are out of range.
```

Comment:
Few things:

- need to separate heading and body
- This is not the correct usage of the term 'safe' (that is reserved for https://doc.rust-lang.org/nomicon/meet-safe-and-unsafe.html). You could say something like "this function does not enforce the invariant that value is within range ...". Instead of 'may be deprecated' can you perhaps say, 'we reserve the right to later enforce the invariant (potentially even by panicing)'.

Reply:
> need to separate heading and body

Thanks, I agree that a separate heading and body is a review good comment. 

> This is not the correct usage of the term 'safe' (that is reserved for https://doc.rust-lang.org/nomicon/meet-safe-and-unsafe.html). You could say something like "this function does not enforce the invariant that value is within range ...".

 I'm hesitant to use the term invariant since it has a very specific meaning in formal language verification as well as mathematics, and we are sort of casually using it here to mean something constant that doesn't change I think.  A loop invariant for example could change within the mechanics of the loop (right?) however by calling it an invariant, we say there are formal assumptions that can be made about the beginning and end of the loop.  Maybe this is also the same reason you object to me using unsafe when I clearly wasn't talking about Rust language semantics.

>  Instead of 'may be deprecated' can you perhaps say, 'we reserve the right to later enforce the invariant (potentially even by panicing)'.

I don't think we should ever change this function to do a check and panic, and instead it should be deprecated and called something else.

What about this:
```
    /// Constructs a new [`SignedAmount`] with satoshi precision and the given number of satoshis.
    ///
    /// This function does not check that [`SignedAmount`] is greater than [`SignedAmount::MIN`]
    /// nor that the value it represents is less than [`SignedAmount::MAX`].  Therefore this
    /// function should be used with caution.  If you wish to represent values that are not in
    /// range, consider using a different type.  Note that the behavior of this function may be
    /// changed in the future to enforce the range of [`SignedAmount`].
```

Thanks for the suggestions and for taking the time to review.

Reply:
>I'm hesitant to use the term invariant since it has a very specific meaning in formal language verification as well as mathematics

Yes, and this meaning applies here. It is an invariant of the `Amount` type that the amount is in range. Just because there's a type-safety loophole via `_unchecked` functions doesn't mean that it's not an invariant. It means that code that uses this loophole to violate the invariant is wrong.

>I don't think we should ever change this function to do a check and panic, and instead it should be deprecated and called something else.

Why? If the user violates an invariant literally anything is allowed to happen, except for unsafety.

Reply:
>  Why? If the user violates an invariant literally anything is allowed to happen, except for unsafety.

To not violate the naming convention of `unchecked_`.  As I understand it, using `unchecked` in the function name implies it does no checks, and if does not checks, how can it panic on out of bounds.,

> Yes, and this meaning applies here. It is an invariant of the Amount type that the amount is in range.

Ok, thanks for the further clarification.  I have only seen invariant commonly used to describe a loop invariant and didn't want to misuse the term when adding docs.  Maybe I will re-word this then using invariant.

Reply:
Updated the text to include discussion of invariant:

```
    /// Constructs a new [`SignedAmount`] with satoshi precision and the given number of satoshis.
    ///
    /// Warning, it's possible to violate the [`SignedAmount`] range invariant.  That is, there is
    /// no check that [`SignedAmount`] is greater than [`SignedAmount::MIN`] nor that the value it
    /// represents is less than [`SignedAmount::MAX`].  If you wish to represent values that are
    /// not in range, consider using a different type.  Note that this function may be modified in
    /// the future to disallow violations of the [`SignedAmount`] range invariant.
```

Reply:
It would be weird for an `_unchecked` method to enforce an invariant by panicking but it's totally allowed (and preferable to unsafety, if somehow our amount range invariant were to become safety-critical) (and IMO preferable to renaming the function).

But the only implication of `_unchecked` is that if you use it wrong your program will be wrong in unspecified and unpredictable ways. If the `_unchecked` method is also marked `unsafe` then your program might even be wrong in a memory-unsafe way. (In many cases there's also an implication that the unchecked method is faster than the equivalent checked one, and we should try to be careful about this.)